File: 백준/Gold/15681. 트리와 쿼리/트리와 쿼리.java
import java.io.*;
import java.util.*;

// 시간 초과 이유: 테케마다 계산을 해버림. 테케에 공통적으로 중복되는 구조가 있으므로 쿼리 전에 모두 계산을 해둔다.

public class Main {

    static int n, r, q;
    static List<Integer> adj[];
    static int parent[];
    static int siz[];

    public static void dfs(int before, int cur) {
        parent[cur] = before;

        for (int nxt : adj[cur]) {
            if (nxt == before) continue;
            dfs(cur, nxt);
        }
    }

    public static void findSubSize(int cur) {
        siz[cur] = 1;
        for (int nxt : adj[cur]) {
            if (nxt == parent[cur]) continue;
            findSubSize(nxt);
            siz[cur] += siz[nxt];
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());
        q = Integer.parseInt(st.nextToken());
        adj = new List[n + 1];
        parent = new int[n + 1];
        siz = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < n - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            adj[u].add(v);
            adj[v].add(u);
        }

        dfs(-1, r);
        findSubSize(r);

        while (q-- > 0) {
            int start = Integer.parseInt(br.readLine());
            System.out.println(siz[start]);
        }
    }
}

File: 백준/Gold/15681. 트리와 쿼리/트리와 쿼리.kt
import java.util.*

/**
 * 쿼리 마다 탐색 -> 시간 초과
 * 한 번에 다 탐색하고 쿼리에 즉답
 */

fun main() = with(System.`in`.bufferedReader()) {
    val (n, r, q) = readLine().split(" ").map { it.toInt() }
    val adj = Array(n + 1) { mutableListOf<Int>() }
    repeat(n - 1) {
        val st = StringTokenizer(readLine())
        val u = st.nextToken().toInt()
        val v = st.nextToken().toInt()
        adj[u].add(v)
        adj[v].add(u)
    }

    val counts = IntArray(n + 1)
    val visited = BooleanArray(n + 1)
    fun countSubTree(cur: Int): Int {
        visited[cur] = true
        for (nxt in adj[cur]) {
            if (visited[nxt]) continue
            counts[cur] += countSubTree(nxt)
        }
        counts[cur]++
        return counts[cur]
    }
    countSubTree(r)

    val sb = StringBuilder()
    repeat(q) {
        sb.append(counts[readLine().toInt()]).append('\n')
    }

    print(sb)
}